instrumentation/grpc-1.5/src/main/java/io/opentelemetry/instrumentation/hypertrace/grpc/v1_5/Http2HeadersInstrumentationModule.java

```diff
@@ -0,0 +1,133 @@
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return Arrays.asList(new NettyUtilsInstrumentation());
+  }
+
+  /**
+   * The server side HTTP2 headers are added in tracing gRPC interceptor. The headers are added to
+   * metadata in {@link GrpcUtils_convertHeaders_Advice}.
+   *
+   * <p>The client side HTTP2 headers are added directly to span in {@link
+   * Utils_convertClientHeaders_Advice}. TODO However it does not work for the first request
+   * https://github.com/hypertrace/javaagent/issues/109#issuecomment-740918018.
+   */
+  class NettyUtilsInstrumentation implements TypeInstrumentation {
+    @Override
+    public ElementMatcher<? super TypeDescription> typeMatcher() {
+      return failSafe(named("io.grpc.netty.Utils"));
+    }
+
+    @Override
+    public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+      Map<ElementMatcher<? super MethodDescription>, String> transformers = new HashMap<>();
+      transformers.put(
+          isMethod().and(named("convertClientHeaders")).and(takesArguments(6)),
+          Utils_convertClientHeaders_Advice.class.getName());
+      transformers.put(
+          isMethod().and(named("convertHeaders")).and(takesArguments(1)),
+          GrpcUtils_convertHeaders_Advice.class.getName());
+      return transformers;
+    }
+  }
+
+  static class Utils_convertClientHeaders_Advice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void exit(
+        @Advice.Argument(1) Object scheme,
+        @Advice.Argument(2) Object defaultPath,
+        @Advice.Argument(3) Object authority,
+        @Advice.Argument(4) Object method) {
+
+      Span currentSpan = Java8BytecodeBridge.currentSpan();
+      if (scheme != null) {
+        currentSpan.setAttribute(
+            HypertraceSemanticAttributes.rpcRequestMetadata(GrpcSemanticAttributes.SCHEME),
+            scheme.toString());
+      }
+      if (defaultPath != null) {
+        currentSpan.setAttribute(
+            HypertraceSemanticAttributes.rpcRequestMetadata(GrpcSemanticAttributes.PATH),
+            defaultPath.toString());
+      }
+      if (authority != null) {
+        currentSpan.setAttribute(
+            HypertraceSemanticAttributes.rpcRequestMetadata(GrpcSemanticAttributes.AUTHORITY),
+            authority.toString());
+      }
+      if (method != null) {
+        currentSpan.setAttribute(
+            HypertraceSemanticAttributes.rpcRequestMetadata(GrpcSemanticAttributes.METHOD),
+            method.toString());
+      }
+    }
+  }
+
+  static class GrpcUtils_convertHeaders_Advice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void exit(
+        @Advice.Argument(0) Http2Headers http2Headers,
+        @Advice.Return(readOnly = false) Metadata metadata) {
+
+      if (http2Headers.authority() != null) {
+        metadata.put(
+            GrpcSemanticAttributes.AUTHORITY_METADATA_KEY, http2Headers.authority().toString());
```

Comment:
we are changing original user metadata object -- need to be very careful if this does not break anything -- especially since this method is called from different flows in code..